### PR TITLE
[sbs] Handle SBS News URLs without IDs

### PR DESCRIPTION
**Pull request by johnsyweb**
### Before submitting this *pull request* I have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/ytdl-org/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/ytdl-org/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/ytdl-org/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [x] Bug fix
- [ ] Improvement
- [x] New extractor
- [ ] New feature

---

#### Before...

```
% ./bin/youtube-dl https://www.sbs.com.au/news/are-the-campaigns-working-voters-speak-out     
[generic] are-the-campaigns-working-voters-speak-out: Requesting header
WARNING: Falling back on generic information extractor.
[generic] are-the-campaigns-working-voters-speak-out: Downloading webpage
[generic] are-the-campaigns-working-voters-speak-out: Extracting information
ERROR: Unsupported URL: https://www.sbs.com.au/news/are-the-campaigns-working-voters-speak-out
```

#### After

```
% ./bin/youtube-dl https://www.sbs.com.au/news/are-the-campaigns-working-voters-speak-out
[SBS] None: Downloading webpage
[SBS] 1520828483782: Downloading JSON metadata
[ThePlatform] Vznr2YGb83mF: Downloading SMIL data
[ThePlatform] Vznr2YGb83mF: Downloading m3u8 information
[ThePlatform] Vznr2YGb83mF: Downloading JSON metadata
[download] Are the campaigns cutting through-Vznr2YGb83mF.mp4 has already been downloaded
[download] 100% of 28.64MiB
```

### CODE REVIEW

File: youtube_dl/extractor/sbs.py [hunk 1]
```diff
@@ -10,15 +12,29 @@
...
+            'uploader': 'SBSC',
+        }
+    }, {
         # Original URL is handled by the generic IE which finds the iframe:
         # http://www.sbs.com.au/thefeed/blog/2014/08/21/dingo-conservation
         'url': 'http://www.sbs.com.au/ondemand/video/single/320403011771/?source=drupal&vertical=thefeed',
         'md5': '3150cf278965eeabb5b4cea1c963fe0a',
         'info_dict': {
-            'id': '320403011771',
+            'id': '_rFBPRPO4pMR',
```
Comment (johnsyweb) [inline youtube_dl/extractor/sbs.py]: Looks like the format has changed.

File: youtube_dl/extractor/sbs.py [hunk 2]
```diff
@@ -36,8 +52,21 @@ class SBSIE(InfoExtractor):
         'only_matching': True,
     }]
 
+    def video_id_from_page_contents(self, url):
+        page_contents = self._download_webpage(url, None)
+        video_id = self._search_regex(r'id="video-(\d+)"', page_contents, 'video id')
+        return video_id
+
+    def video_id(self, url):
+        ID_BEARING_URL = r'https?://(?:www\.)?sbs\.com\.au/(?:ondemand|news)/video/(?:single/)?(?P<id>[0-9]+)'
```
Comment (dstftw) [inline youtube_dl/extractor/sbs.py]: Code duplication.

File: youtube_dl/extractor/sbs.py [hunk 3]
```diff
@@ -36,8 +52,21 @@ class SBSIE(InfoExtractor):
         'only_matching': True,
     }]
 
+    def video_id_from_page_contents(self, url):
+        page_contents = self._download_webpage(url, None)
```
Comment (dstftw) [inline youtube_dl/extractor/sbs.py]: No such id `None`.

File: youtube_dl/extractor/sbs.py [hunk 4]
```diff
@@ -10,15 +12,29 @@
 
 class SBSIE(InfoExtractor):
     IE_DESC = 'sbs.com.au'
-    _VALID_URL = r'https?://(?:www\.)?sbs\.com\.au/(?:ondemand|news)/video/(?:single/)?(?P<id>[0-9]+)'
+    _VALID_URL = r'https?://(?:www\.)?sbs\.com\.au/(?:ondemand|news)/(?:video/)?(?:single/)?([0-9]+|[0-9a-z-]+)'
```
Comment (dstftw) [inline youtube_dl/extractor/sbs.py]: Must be a separate extractor delegating to SBS.

File: youtube_dl/extractor/sbs.py (hunk 3 above)
Comment (johnsyweb) [reply youtube_dl/extractor/sbs.py]: `id` doesn't seem to be used in this flow.

File: youtube_dl/extractor/sbs.py (hunk 4 above)
Comment (johnsyweb) [reply youtube_dl/extractor/sbs.py]: Thanks for your review, @dstftw. I agree, this does feel like a separate extractor. Addressed in 744df7ab325c6ee28c5a877002adfba91f0e7a36 . Let me know if you need anything further.

**Comment by johnsyweb** [comment]
@dstftw : Do my changes address all of your concerns?

**Comment by johnsyweb** [comment]
@dstftw Can we please merge or close this?

File: youtube_dl/extractor/sbs.py [hunk 5]
```diff
@@ -64,3 +64,35 @@ def _real_extract(self, url):
...
+        },
+    }, {
+        'url': 'https://www.sbs.com.au/news/sbs-world-news-bulletin-may-11',
+        'only_matching': True,
+    }]
+
+    def _video_id(self, url):
+        slug = self._match_id(url)
+        page_contents = self._download_webpage(url, slug)
+        return self._search_regex(r'id="video-(\d+)"', page_contents, 'video id')
```
Comment (dstftw) [inline youtube_dl/extractor/sbs.py:98]: As already pointed out: you must delegate to `SBSIE` extractor not inherit from it.

File: youtube_dl/extractor/sbs.py [hunk 6]
```diff
@@ -64,3 +64,35 @@ def _real_extract(self, url):
             'id': video_id,
             'url': smuggle_url(self._proto_relative_url(theplatform_url), {'force_smil_url': True}),
         }
+
+    def _video_id(self, url):
+        return self._match_id(url)
+
+
+class SBSNewsIE(SBSIE):
+    _VALID_URL = r'^https?://(?:www\.)?sbs\.com\.au/news/(?P<id>[0-9a-z-]+)$'
```
Comment (dstftw) [inline youtube_dl/extractor/sbs.py:73]: No `^` and `$`. Override `suitable`.

File: youtube_dl/extractor/sbs.py [hunk 7]
```diff
@@ -64,3 +64,35 @@ def _real_extract(self, url):
...
+            'id': 'Vznr2YGb83mF',
+            'ext': 'mp4',
+            'title': 'Are the campaigns cutting through?',
+            'description': 'md5:d41d8cd98f00b204e9800998ecf8427e',
+            'thumbnail': r're:http://.*\.jpg',
+            'duration': 146,
+            'timestamp': 1557552900,
+            'upload_date': '20190511',
+            'uploader': 'SBSC',
+        },
```
Comment (dstftw) [inline youtube_dl/extractor/sbs.py:89]: This is useless if you use `only_matching`.

Comment (johnsyweb) [reply youtube_dl/extractor/sbs.py]: Addressed in 5e04505.

File: youtube_dl/extractor/sbs.py (hunk 6 above)
Comment (johnsyweb) [reply youtube_dl/extractor/sbs.py]: Addressed in fd90cab  and 685ed27.

File: youtube_dl/extractor/sbs.py (hunk 5 above)
Comment (johnsyweb) [reply youtube_dl/extractor/sbs.py:96]: There's no obvious delegation pattern elsewhere in this repository for a new contributor to follow.  

In 5d6e557, I extract SBSBaseIE with common _real_extract(), which follows established BaseIE pattern and keeps the code readable. Is this acceptable to you, @dstftw?

Comment (dstftw) [reply youtube_dl/extractor/sbs.py:96]: Not acceptable. Search `url_result`.

Comment (johnsyweb) [reply youtube_dl/extractor/sbs.py:96]: I'm sorry, @dstftw, could you please explain to me _why_ this solution is not acceptable? Or perhaps explain how the 400+ search results for the `url_result` utility method is relevant to implementing a common `_real_extract()` method? I'm really not clear on what it is I need to do to get this PR into the shape you want for it to be merged.

**Comment by johnsyweb** [comment]
This PR has been open for over a year and the feedback on it hasn't helped me get it in a state to merge. 

I sought some help in the IRC channel but none was forthcoming, so I think I'll just close it.

**Pull request closed without merging**